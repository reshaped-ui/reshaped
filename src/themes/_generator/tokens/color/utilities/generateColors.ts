import type { ThemeDefinition } from "themes/_generator/tokens/types";
import {
	Token,
	HexColor,
	OklchToken,
	type Name as TokenName,
	type ColorValue,
	Hue,
	OklchColor,
} from "../color.types";
import { hexToOklch, tokenToOklchToken } from "./convert";

const parseColor = (color: HexColor | Token) => {
	const isString = typeof color === "string";

	if (!isString && "oklch" in color && color.oklch) {
		return {
			light: color.oklch,
			dark: color.oklchDark,
		};
	}

	const hexLight = isString ? color : color.hex;
	const hexDark = isString ? undefined : color.hexDark;

	const light = hexToOklch(hexLight);
	const dark = hexDark ? hexToOklch(hexDark) : undefined;

	if (!light) {
		throw new Error(`[Reshaped] Failed when parsing color: ${JSON.stringify(color)}`);
	}

	return { light, dark };
};

const getDarkModeColor = (lch: OklchColor) => {
	const { l, c } = lch;

	// Top boundary at which background colors are perceived as vibrant
	const vibrancyThreshold = 0.25;
	// A linear modifier of how gray a color is, gray = 1, vibrant = 0
	const neutralFactor = Math.min(1, Math.max(0, (vibrancyThreshold - c) / vibrancyThreshold));
	// Non-linear modification to make gray color inversion progress faster
	const easedVibrancy = 1 - Math.pow(neutralFactor, 2);

	const invertedL = 1 - l;
	// Take plain inversion and increase it back by vibrancy modifier
	// Vibrant colors should get back to almost the original value
	// while non-briant colors should lean towards full inversion

	// Example:
	// Black should become white in dark mode
	// Light gray should become dark gray
	// Vibrant blue should stay vibrant blue
	const darkL = invertedL + (l - invertedL) * easedVibrancy;

	// Make sure dark mode value doesn't become too dark and is still visible on dark page background
	const clampedDarkL = Math.max(0.36, darkL);

	return { ...lch, c: c * 0.9, l: clampedDarkL };
};

const generateColorValues = (
	key: string,
	token: ColorValue
): Partial<Record<TokenName, OklchToken>> => {
	const { light: bg, dark } = parseColor(token);
	const capitalizedKey = key.charAt(0).toUpperCase() + key.slice(1);
	const neutral = key === "neutral";
	const warning = key === "warning";

	const bgDark = dark || getDarkModeColor(bg);
	const bgFaded = {
		...bg,
		l: 0.97,
		c: neutral ? 0.005 : warning ? 0.04 : 0.02,
	};
	const bgFadedDark = {
		...bgDark,
		l: 0.25,
		// For primary color with low chroma, we still have to make sure it stays low
		c: neutral ? 0.01 : bgDark.c / 4,
	};

	const fg = neutral ? { ...bg, l: 0.2 } : { ...bg, l: 0.5 };
	const fgDark = neutral ? { ...bgDark, l: 0.96 } : { ...bgDark, l: 0.75, c: bg.c * 0.85 };

	const bd = neutral ? { ...bg, l: 0, alpha: 0.12 } : { ...bg, l: bg.l - 0.08 };
	const bdDark = neutral ? { ...bgDark, l: 1, alpha: 0.16 } : { ...bgDark, l: bgDark.l + 0.08 };
	const bdFaded = neutral
		? { ...bgFaded, l: 0, alpha: 0.08 }
		: { ...bgFaded, l: bgFaded.l - 0.05, c: bgFaded.c + 0.01 };
	const bdFadedDark = neutral
		? { ...bgFadedDark, l: 1, alpha: 0.08 }
		: { ...bgFadedDark, l: bgFadedDark.l + 0.1 };

	const output = {
		[`background${capitalizedKey}`]: {
			oklch: bg,
			oklchDark: bgDark,
		},
		[`background${capitalizedKey}Faded`]: {
			oklch: bgFaded,
			oklchDark: bgFadedDark,
		},
		[`border${capitalizedKey}`]: {
			oklch: bd,
			oklchDark: bdDark,
		},
		[`border${capitalizedKey}Faded`]: {
			oklch: bdFaded,
			oklchDark: bdFadedDark,
		},
		[`foreground${capitalizedKey}`]: {
			oklch: fg,
			oklchDark: fgDark,
		},
	};

	if (neutral) {
		output[`foreground${capitalizedKey}Faded`] = {
			oklch: { ...fg, l: fg.l + 0.25 },
			oklchDark: { ...fgDark, l: fgDark.l - 0.15 },
		};
		output[`backgroundDisabled`] = {
			oklch: { ...bg, l: 0.95, c: 0 },
			oklchDark: { ...bgDark, l: 0.28, c: 0 },
		};
		output[`backgroundDisabledFaded`] = {
			oklch: { ...bg, l: 0.98, c: 0 },
			oklchDark: { ...bgDark, l: 0.23, c: 0 },
		};
		output[`borderDisabled`] = {
			oklch: { ...bd, alpha: 0.06 },
			oklchDark: { ...bgDark, l: 0.28, c: 0 },
		};
		output[`foregroundDisabled`] = {
			oklch: { ...fg, l: 0.84, c: 0 },
			oklchDark: { ...fgDark, l: 0.4, c: 0 },
		};
		output[`backgroundElevationBase`] = {
			oklch: { ...bg, l: 1, c: 0 },
			oklchDark: { ...bgDark, l: 0.2, c: 0 },
		};
		output[`backgroundElevationRaised`] = {
			oklch: { ...bg, l: 1, c: 0 },
			oklchDark: { ...bgDark, l: 0.21 },
		};
		output[`backgroundElevationOverlay`] = {
			oklch: { ...bg, l: 1, c: 0 },
			oklchDark: { ...bgDark, l: 0.22, c: 0 },
		};
		output[`backgroundPage`] = {
			oklch: { ...bg, l: 1, c: 0 },
			oklchDark: { ...bgDark, l: 0.16, c: 0 },
		};
		output[`backgroundPageFaded`] = {
			oklch: { ...bg, l: 0.97, c: 0 },
			oklchDark: { ...bgDark, l: 0.18, c: 0 },
		};
	}

	return output;
};

const getOklchToken = (color: ColorValue) => {
	const token = typeof color === "string" ? { hex: color } : color;
	return tokenToOklchToken(token);
};

const generateColors = (args: Partial<Record<Hue, ColorValue>> = {}) => {
	const {
		primary = "#1469ff",
		critical = "#e22c2c",
		warning = "#facc15",
		positive = "#118850",
		neutral = "#e3e4e8",
		brand,
	} = args;

	return {
		...generateColorValues("primary", getOklchToken(primary)),
		...generateColorValues("critical", getOklchToken(critical)),
		...generateColorValues("warning", getOklchToken(warning)),
		...generateColorValues("positive", getOklchToken(positive)),
		...generateColorValues("neutral", getOklchToken(neutral)),
		brand: getOklchToken(brand || primary),
		white: getOklchToken("#ffffff"),
		black: getOklchToken("#000000"),
	} as ThemeDefinition["color"];
};

export default generateColors;
