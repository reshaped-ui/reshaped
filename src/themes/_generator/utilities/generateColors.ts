import type { ThemeDefinition } from "../tokens/types";
import type { Color } from "../types";
import {
	parseHex,
	convertOklabToOkhsl,
	convertLrgbToOklab,
	convertRgbToLrgb,
	convertOkhslToOklab,
	convertOklabToLrgb,
	convertLrgbToRgb,
	type Okhsl,
	type Rgb,
	serializeHex,
	serializeHex8,
} from "culori/fn";

const hexToOkHsl = (hex: string) => {
	const rgb = parseHex(hex) as Rgb;

	if (!rgb) throw new Error(`Can\'t generate rgb from ${hex} color`);

	const lrgb = convertRgbToLrgb(rgb);
	const oklab = convertLrgbToOklab(lrgb);
	const okhsl = convertOklabToOkhsl(oklab);

	return okhsl;
};

const okhslToHex = (okhsl: Okhsl) => {
	const oklab = convertOkhslToOklab(okhsl);
	const lrgb = convertOklabToLrgb(oklab);
	const rgb = convertLrgbToRgb(lrgb);
	return !okhsl.alpha || okhsl.alpha === 1 ? serializeHex(rgb) : serializeHex8(rgb);
};

const getDarkModeColor = (hsl: Okhsl) => {
	const { l } = hsl;

	const mid = 0.5;
	const lDelta = 1 - (l - mid) / (1 - mid);

	return { ...hsl, l: l < mid ? l : 1 - l * lDelta };
};

const generateColorValues = (args: { key: string; hex: string; hexDark?: string }) => {
	const { key, hex, hexDark } = args;
	const capitalizedKey = key.charAt(0).toUpperCase() + key.slice(1);

	const okhsl = hexToOkHsl(hex);

	/**
	 * Calculate hue perceptual lighness (yellow, cyan, )
	 */
	const hueStep = 120;
	const normalizedHue = ((okhsl.h! + 150) % 360) / 360;
	const hueLightness = Math.cos((2 * Math.PI) / hueStep) * normalizedHue;

	const okhslDark =
		key === "neutral"
			? { ...okhsl, l: 1 - okhsl.l + 0.16, s: 0.1 }
			: (hexDark && hexToOkHsl(hexDark)) || getDarkModeColor(okhsl);

	/**
	 * Background
	 */
	const bgHex = hex;
	const bgHexDark = okhslToHex(okhslDark);

	const bgFadedHsl = { ...okhsl, l: 0.94 + 0.045 * hueLightness };
	const bgFadedHslDark = { ...okhslDark, l: 0.16, s: okhslDark.s / 2 };
	const bgFadedHex = okhslToHex(bgFadedHsl);
	const bgFadedHexDark = okhslToHex(bgFadedHslDark);

	const bgHighlightedHex = okhslToHex({ ...okhsl, l: okhsl.l - 0.04 });
	const bgHighlightedHexDark = okhslToHex({ ...okhslDark, l: okhslDark.l + 0.04 });

	/**
	 * Foreground
	 */

	const fgOkhsl = key === "neutral" ? { ...okhsl, l: 0.2 } : { ...okhsl, l: 0.45 };
	// Lighter colors need smaller lightness increase for fg colors
	const fgDarkLDelta = 0.16 / (1 + Math.max(0, okhslDark.l - 0.5));
	const fgOkhslDark =
		key === "neutral"
			? { ...okhsl, l: 0.96 }
			: { ...okhslDark, l: okhslDark.l + fgDarkLDelta, s: 0.7 };
	const fgHex = okhslToHex(fgOkhsl);
	const fgHexDark = okhslToHex(fgOkhslDark);

	/**
	 * Border
	 */
	const bdHex = key === "neutral" ? okhslToHex({ ...fgOkhsl, l: 0, alpha: 0.12 }) : fgHex;
	const bdHexDark =
		key === "neutral" ? okhslToHex({ ...fgOkhslDark, l: 1, alpha: 0.14 }) : fgHexDark;

	const bdFadedHex =
		key === "neutral"
			? okhslToHex({
					...bgFadedHsl,
					l: 0,
					alpha: 0.06,
				})
			: okhslToHex({
					...bgFadedHsl,
					s: 0.6 - 0.2 * hueLightness,
					l: bgFadedHsl.l - 0.05,
				});
	const bdFadedHexDark =
		key === "neutral"
			? okhslToHex({
					...bgFadedHslDark,
					l: 1,
					alpha: 0.09,
				})
			: okhslToHex({
					...bgFadedHslDark,
					l: bgFadedHslDark.l + 0.08,
				});

	const output = {
		[`background${capitalizedKey}`]: {
			hex: bgHex,
			hexDark: bgHexDark,
		},
		[`background${capitalizedKey}Faded`]: {
			hex: bgFadedHex,
			hexDark: bgFadedHexDark,
		},
		[`background${capitalizedKey}Highlighted`]: {
			hex: bgHighlightedHex,
			hexDark: bgHighlightedHexDark,
		},
		[`border${capitalizedKey}`]: {
			hex: bdHex,
			hexDark: bdHexDark,
		},
		[`border${capitalizedKey}Faded`]: {
			hex: bdFadedHex,
			hexDark: bdFadedHexDark,
		},
		[`foreground${capitalizedKey}`]: {
			hex: fgHex,
			hexDark: fgHexDark,
		},
	};

	if (key === "neutral") {
		output[`foreground${capitalizedKey}Faded`] = {
			hex: okhslToHex({ ...fgOkhsl, l: fgOkhsl.l + 0.25 }),
			hexDark: okhslToHex({ ...fgOkhslDark, l: fgOkhslDark.l - 0.15 }),
		};
		output[`backgroundDisabled`] = {
			hex: okhslToHex({ ...okhsl, s: 0.04, l: 0.94 }),
			hexDark: okhslToHex({ ...okhslDark, s: 0.04, l: 0.16 }),
		};
		output[`backgroundDisabledFaded`] = {
			hex: okhslToHex({ ...okhsl, l: 0.97, s: 0.04 }),
			hexDark: okhslToHex({ ...okhslDark, l: 0.12, s: 0.04 }),
		};
		output[`borderDisabled`] = {
			hex: okhslToHex({ ...okhsl, l: 0.9, s: 0.04 }),
			hexDark: okhslToHex({ ...okhslDark, l: 0.17, s: 0.04 }),
		};
		output[`foregroundDisabled`] = {
			hex: okhslToHex({ ...okhsl, l: 0.82, s: 0.04 }),
			hexDark: okhslToHex({ ...okhslDark, l: 0.32, s: 0.04 }),
		};
		output[`backgroundElevationBase`] = {
			hex: "#ffffff",
			hexDark: okhslToHex({ ...okhslDark, s: 0.03, l: 0.1 }),
		};
		output[`backgroundElevationRaised`] = {
			hex: "#ffffff",
			hexDark: okhslToHex({ ...okhslDark, s: 0.04, l: 0.11 }),
		};
		output[`backgroundElevationOverlay`] = {
			hex: "#ffffff",
			hexDark: okhslToHex({ ...okhslDark, s: 0.04, l: 0.13 }),
		};
		output[`backgroundPage`] = {
			hex: "#ffffff",
			hexDark: okhslToHex({ ...okhslDark, l: 0.07, s: 0.02 }),
		};
		output[`backgroundPageFaded`] = {
			hex: okhslToHex({ ...okhsl, l: 0.985 }),
			hexDark: okhslToHex({ ...okhslDark, l: 0.09, s: 0.02 }),
		};
	}

	return output;
};

const validateHexColor = (color: string) => {
	const hexColorRegex = /^#([A-Fa-f0-9]{3}){2}$/;

	if (!hexColorRegex.test(color)) {
		throw new Error(`Invalid hex color: ${color}`);
	}

	return color;
};

const generate = (
	args: {
		primary?: Color;
		critical?: Color;
		warning?: Color;
		positive?: Color;
		neutral?: Color;
		brand?: string;
	} = {}
) => {
	const {
		primary = "#2563eb",
		critical = "#e22c2c",
		warning = "#facc15",
		positive = "#118850",
		neutral = "#e3e4e8",
		brand,
	} = args;

	const generateFor = (key: string, color: Color) => {
		return generateColorValues({
			key,
			hex: validateHexColor(typeof color === "string" ? color : color.hex),
			hexDark: typeof color !== "string" ? validateHexColor(color.hexDark) : undefined,
		});
	};

	return {
		...generateFor("primary", primary),
		...generateFor("critical", critical),
		...generateFor("warning", warning),
		...generateFor("positive", positive),
		...generateFor("neutral", neutral),
		brand: { hex: brand || primary },
		white: { hex: "#ffffff" },
		black: { hex: "#000000" },
	} as ThemeDefinition["color"];
};

export default generate;
