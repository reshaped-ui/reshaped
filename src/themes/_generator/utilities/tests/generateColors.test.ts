import generateColors from "../generateColors";

describe("Themes/Utilties/generateColors", () => {
	test("generate the default theme", () => {
		const colors = generateColors();

		expect(colors).toMatchSnapshot();
	});

	test("applies color overrides", () => {
		const colors = generateColors({
			primary: "#2563eb",
			critical: "#dc2626",
			positive: "#16a34a",
		});

		expect(colors).toMatchSnapshot();
	});
});
