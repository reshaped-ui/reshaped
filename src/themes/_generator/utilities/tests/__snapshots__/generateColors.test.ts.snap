// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Themes/Utilties/generateColors applies color overrides 1`] = `
{
  "backgroundCritical": {
    "hex": "#dc2626",
    "hexDark": "#c02929",
  },
  "backgroundCriticalFaded": {
    "hex": "#fceeee",
    "hexDark": "#3e1f1f",
  },
  "backgroundCriticalHighlighted": {
    "hex": "#ca2222",
    "hexDark": "#d22e2e",
  },
  "backgroundDisabled": {
    "hex": "#eceef2",
    "hexDark": "#1e212a",
  },
  "backgroundDisabledFaded": {
    "hex": "#f5f6f9",
    "hexDark": "#171921",
  },
  "backgroundElevationBase": {
    "hex": "#ffffff",
    "hexDark": "#15171e",
  },
  "backgroundElevationOverlay": {
    "hex": "#ffffff",
    "hexDark": "#1c1f28",
  },
  "backgroundElevationRaised": {
    "hex": "#ffffff",
    "hexDark": "#191b23",
  },
  "backgroundNeutral": {
    "hex": "#dfe2ea",
    "hexDark": "#494f60",
  },
  "backgroundNeutralFaded": {
    "hex": "#f1f2f6",
    "hexDark": "#222835",
  },
  "backgroundNeutralHighlighted": {
    "hex": "#d2d7e2",
    "hexDark": "#52586b",
  },
  "backgroundPage": {
    "hex": "#ffffff",
    "hexDark": "#0f1116",
  },
  "backgroundPageFaded": {
    "hex": "#f9f9fb",
    "hexDark": "#111319",
  },
  "backgroundPositive": {
    "hex": "#16a34a",
    "hexDark": "#1a8d44",
  },
  "backgroundPositiveFaded": {
    "hex": "#edfdf3",
    "hexDark": "#1f2b21",
  },
  "backgroundPositiveHighlighted": {
    "hex": "#149744",
    "hexDark": "#1d994a",
  },
  "backgroundPrimary": {
    "hex": "#2563eb",
    "hexDark": "#1e5bdf",
  },
  "backgroundPrimaryFaded": {
    "hex": "#ecf1fd",
    "hexDark": "#23273b",
  },
  "backgroundPrimaryHighlighted": {
    "hex": "#215ad7",
    "hexDark": "#2264f3",
  },
  "backgroundWarning": {
    "hex": "#facc15",
    "hexDark": "#f1c512",
  },
  "backgroundWarningFaded": {
    "hex": "#fffae9",
    "hexDark": "#2c271f",
  },
  "backgroundWarningHighlighted": {
    "hex": "#edc113",
    "hexDark": "#fed014",
  },
  "black": {
    "hex": "#000000",
  },
  "borderCritical": {
    "hex": "#bb1f1f",
    "hexDark": "#f16c6c",
  },
  "borderCriticalFaded": {
    "hex": "#f7dcdc",
    "hexDark": "#582929",
  },
  "borderDisabled": {
    "hex": "#dfe2ea",
    "hexDark": "#262a34",
  },
  "borderNeutral": {
    "hex": "#b3bbce",
    "hexDark": "#4c5264",
  },
  "borderNeutralFaded": {
    "hex": "#dddfe7",
    "hexDark": "#2d374c",
  },
  "borderPositive": {
    "hex": "#10853b",
    "hexDark": "#22ac54",
  },
  "borderPositiveFaded": {
    "hex": "#cff3dd",
    "hexDark": "#283b2c",
  },
  "borderPrimary": {
    "hex": "#1c50c1",
    "hexDark": "#758ff9",
  },
  "borderPrimaryFaded": {
    "hex": "#d9e3f9",
    "hexDark": "#2e3555",
  },
  "borderWarning": {
    "hex": "#cfa90f",
    "hexDark": "#b4920a",
  },
  "borderWarningFaded": {
    "hex": "#faedbb",
    "hexDark": "#3d3628",
  },
  "brand": {
    "hex": "#2563eb",
  },
  "foregroundCritical": {
    "hex": "#c62121",
    "hexDark": "#f56868",
  },
  "foregroundDisabled": {
    "hex": "#c6ccda",
    "hexDark": "#434959",
  },
  "foregroundNeutral": {
    "hex": "#14181f",
    "hexDark": "#eff1f5",
  },
  "foregroundNeutralFaded": {
    "hex": "#5b667e",
    "hexDark": "#c0c6d6",
  },
  "foregroundPositive": {
    "hex": "#0d7533",
    "hexDark": "#18ac4f",
  },
  "foregroundPrimary": {
    "hex": "#225cdc",
    "hexDark": "#7590f7",
  },
  "foregroundWarning": {
    "hex": "#7b6305",
    "hexDark": "#b4920c",
  },
  "white": {
    "hex": "#ffffff",
  },
}
`;

exports[`Themes/Utilties/generateColors generate the default theme 1`] = `
{
  "backgroundCritical": {
    "hex": "#e22c2c",
    "hexDark": "#d02626",
  },
  "backgroundCriticalFaded": {
    "hex": "#fdeded",
    "hexDark": "#3e1f1f",
  },
  "backgroundCriticalHighlighted": {
    "hex": "#d02828",
    "hexDark": "#e22a2a",
  },
  "backgroundDisabled": {
    "hex": "#eceef2",
    "hexDark": "#1e212a",
  },
  "backgroundDisabledFaded": {
    "hex": "#f5f6f9",
    "hexDark": "#171921",
  },
  "backgroundElevationBase": {
    "hex": "#ffffff",
    "hexDark": "#15171e",
  },
  "backgroundElevationOverlay": {
    "hex": "#ffffff",
    "hexDark": "#1c1f28",
  },
  "backgroundElevationRaised": {
    "hex": "#ffffff",
    "hexDark": "#191b23",
  },
  "backgroundNeutral": {
    "hex": "#dfe2ea",
    "hexDark": "#494f60",
  },
  "backgroundNeutralFaded": {
    "hex": "#f1f2f6",
    "hexDark": "#222835",
  },
  "backgroundNeutralHighlighted": {
    "hex": "#d2d7e2",
    "hexDark": "#52586b",
  },
  "backgroundPage": {
    "hex": "#ffffff",
    "hexDark": "#0f1116",
  },
  "backgroundPageFaded": {
    "hex": "#f9f9fb",
    "hexDark": "#111319",
  },
  "backgroundPositive": {
    "hex": "#118850",
    "hexDark": "#14784a",
  },
  "backgroundPositiveFaded": {
    "hex": "#edfdf5",
    "hexDark": "#1f2a23",
  },
  "backgroundPositiveHighlighted": {
    "hex": "#0f7d49",
    "hexDark": "#178351",
  },
  "backgroundPrimary": {
    "hex": "#5a58f2",
    "hexDark": "#524fea",
  },
  "backgroundPrimaryFaded": {
    "hex": "#ecebfe",
    "hexDark": "#252544",
  },
  "backgroundPrimaryHighlighted": {
    "hex": "#4d4af0",
    "hexDark": "#5f5ced",
  },
  "backgroundWarning": {
    "hex": "#facc15",
    "hexDark": "#f1c512",
  },
  "backgroundWarningFaded": {
    "hex": "#fffae9",
    "hexDark": "#2c271f",
  },
  "backgroundWarningHighlighted": {
    "hex": "#edc113",
    "hexDark": "#fed014",
  },
  "black": {
    "hex": "#000000",
  },
  "borderCritical": {
    "hex": "#bf2424",
    "hexDark": "#f46969",
  },
  "borderCriticalFaded": {
    "hex": "#fadbdb",
    "hexDark": "#582929",
  },
  "borderDisabled": {
    "hex": "#dfe2ea",
    "hexDark": "#262a34",
  },
  "borderNeutral": {
    "hex": "#b3bbce",
    "hexDark": "#4c5264",
  },
  "borderNeutralFaded": {
    "hex": "#dddfe7",
    "hexDark": "#2d374c",
  },
  "borderPositive": {
    "hex": "#0c6e40",
    "hexDark": "#21ab6b",
  },
  "borderPositiveFaded": {
    "hex": "#d0f3e2",
    "hexDark": "#293b2f",
  },
  "borderPrimary": {
    "hex": "#3b38ed",
    "hexDark": "#8c8bf3",
  },
  "borderPrimaryFaded": {
    "hex": "#dedcfb",
    "hexDark": "#323164",
  },
  "borderWarning": {
    "hex": "#cfa90f",
    "hexDark": "#b4920a",
  },
  "borderWarningFaded": {
    "hex": "#faedbb",
    "hexDark": "#3d3628",
  },
  "brand": {
    "hex": "#5a58f2",
  },
  "foregroundCritical": {
    "hex": "#c42525",
    "hexDark": "#f36a6a",
  },
  "foregroundDisabled": {
    "hex": "#c6ccda",
    "hexDark": "#434959",
  },
  "foregroundNeutral": {
    "hex": "#14181f",
    "hexDark": "#eff1f5",
  },
  "foregroundNeutralFaded": {
    "hex": "#5b667e",
    "hexDark": "#c0c6d6",
  },
  "foregroundPositive": {
    "hex": "#0d7544",
    "hexDark": "#18ab66",
  },
  "foregroundPrimary": {
    "hex": "#4f4cf0",
    "hexDark": "#8b8af7",
  },
  "foregroundWarning": {
    "hex": "#7b6305",
    "hexDark": "#b4920c",
  },
  "white": {
    "hex": "#ffffff",
  },
}
`;
